Api/Functions/SlackIncoming.cs

```diff
@@ -0,0 +1,99 @@
+﻿using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Azure.WebJobs;
+using Microsoft.Azure.WebJobs.Extensions.Http;
+using Microsoft.AspNetCore.Http;
+using System.IO;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json.Linq;
+using System.Net.Http;
+using Merq;
+using NosAyudamos.Events;
+using System.Globalization;
+
+namespace NosAyudamos.Functions
+{
+    class SlackIncoming
+    {
+        const string ApiUrl = "https://slack.com/api/";
+
+        readonly ISerializer serializer;
+        readonly IEnvironment environment;
+        readonly IEventStream events;
+        readonly ILanguageUnderstanding language;
+        readonly HttpClient http;
+        readonly ILogger<SlackIncoming> logger;
+
+        public SlackIncoming(ISerializer serializer, IEnvironment environment, IEventStream events, ILanguageUnderstanding language, HttpClient http, ILogger<SlackIncoming> logger) =>
+            (this.serializer, this.environment, this.events, this.language, this.http, this.logger) = (serializer, environment, events, language, http, logger);
+
+        [FunctionName("slack_interaction")]
+        public async Task<IActionResult> InteractionAsync(
+            [HttpTrigger(AuthorizationLevel.Anonymous, "post", Route = "slack/interaction")] HttpRequest req)
+        {
+            using var reader = new StreamReader(req.Body);
+            var payload = await reader.ReadToEndAsync();
+            dynamic json = serializer.Deserialize<JObject>(System.Net.WebUtility.UrlDecode(payload.Substring(8)));
+
+            if (json.challenge != null)
+            {
+                return new OkObjectResult(json.challenge);
+            }
+
+            string? intent = json.actions?[0]?.value;
+            string? message = json.message?.blocks?[2]?.text?.text;
+            if (message != null && message.Trim().StartsWith("&gt;", System.StringComparison.Ordinal))
+                message = message.Trim().Substring(4);
+
+            if (message != null)
+                message = message.Trim();
+
+            // TODO: uncomment when new API is merged.
+            //await language.AddUtteranceAsync(message, intent);
+            logger.LogInformation("Training intent '{intent}' with new phrase: {phrase}", intent, message);
+
+            return new OkResult();
+        }
+
+        [FunctionName("slack_incoming")]
+        public async Task<IActionResult> ReceiveAsync(
```

Comment:
Whats the purpose of this azure function?

Reply:
It's the webhook that receives incoming messages from slack (i.e. you responding a thread from the unknown message)